### PR TITLE
Make useFsEvents as default strategy for the watching

### PR DESCRIPTION
**Pull request by sheetalkamat**
Now that we have gotten better at handling inode movement etc and many people have tried using `useFsEvents` as default for watching files, make it default. This should give more agile callbacks compared to polling especially if it involves lots of files. It should also be better on CPU.

One catch on systems which have limits on fs watchers like linux, it would still fallback to polling but that was anyways done for the recursive directory watching so files is no different, but quantity varies so something to keep in mind.

This change also ensures that fsWatches are cached and reused with multiple callbacks instead of creating different fs watcher.

Most of the changes are refactoring on better baseline test infra update
[10a7b4e](https://github.com/microsoft/TypeScript/pull/50366/commits/10a7b4e252f2d6554895a923c48435d7211d30a7) actually has change to make fsEvents as default for file watching
[61a99bf](https://github.com/microsoft/TypeScript/pull/50366/commits/61a99bf49426ddd6c6e883a8856f161afa47b4a3) has change to cache the watches per name even when using fsEvents

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
I wonder if you can see a change TSServer perf tests.

@typescript-bot perf test this

File: src/compiler/sys.ts [hunk 1]
```diff
@@ -418,50 +418,41 @@ namespace ts {
...
-                        pollingInterval,
-                        options
-                    ),
-                    refCount: 1
-                });
+        return {
+            close: () => {
+                const watcher = cache.get(path);
+                if (!watcher) return;
+                if (!orderedRemoveItem(watcher.callbacks, callback) || watcher.callbacks.length) return;
```
Comment (amcasey) [inline src/compiler/sys.ts:452]: We're using `orderedRemoveItem` to guarantee that the callbacks are called in the order in which they were attached?

File: src/compiler/sys.ts [hunk 2]
```diff
@@ -418,50 +418,41 @@ namespace ts {
...
-                        ),
-                        pollingInterval,
-                        options
-                    ),
-                    refCount: 1
-                });
+        return {
+            close: () => {
+                const watcher = cache.get(path);
+                if (!watcher) return;
```
Comment (amcasey) [inline src/compiler/sys.ts:451]: Can this happen?  Why would it not be in the cache?

File: src/compiler/sys.ts (hunk 1 above)
Comment (amcasey) [inline src/compiler/sys.ts:452]: Can the call to `orderedRemoveItem` actually return false?

File: src/compiler/sys.ts [hunk 3]
```diff
@@ -1064,13 +1057,39 @@ namespace ts {
...
+            return createSingleWatcherPerName(
+                recursive ? fsWatchesRecursive : fsWatches,
+                useCaseSensitiveFileNames,
+                fileOrDirectory,
+                callback,
+                cb => fsWatchHandlingPresence(fileOrDirectory, entryKind, cb, recursive, fallbackPollingInterval, fallbackOptions),
+            );
+        }
+
+        function fsWatchHandlingPresence(
```
Comment (amcasey) [inline src/compiler/sys.ts:1086]: What does "handling presence" mean?

File: src/compiler/sys.ts [hunk 4]
```diff
@@ -418,50 +418,41 @@ namespace ts {
...
+        const path = toCanonicalFileName(name);
+        const existing = cache.get(path);
+        if (existing) {
+            existing.callbacks.push(callback);
+        }
+        else {
+            cache.set(path, {
+                watcher: createWatcher((
+                    // Cant infer types correctly so lets satisfy checker
+                    (param1: any, param2: never, param3: any) => cache.get(path)?.callbacks.slice().forEach(cb => cb(param1, param2, param3))
```
Comment (amcasey) [inline src/compiler/sys.ts:442]: Why slice?  Is that guarding against modification during iteration or something?

Comment (amcasey) [inline src/compiler/sys.ts:442]: What does the `?` guard against?  Is that for a file change that happens after the watcher is closed or something?

**Review by amcasey — COMMENTED** [review]
I had trouble following some parts.

File: src/compiler/sys.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/compiler/sys.ts:452]: Yes

Comment (sheetalkamat) [reply src/compiler/sys.ts:452]: Some exceptions thrown while setting watch or closing watch can cause this and its better to write defensive code than get unnecessary asserts i think ..

File: src/compiler/sys.ts (hunk 4 above)
Comment (sheetalkamat) [reply src/compiler/sys.ts:442]: Yes.

File: src/compiler/sys.ts (hunk 2 above)
Comment (sheetalkamat) [reply src/compiler/sys.ts:451]: Same as before.. defensive to handle the exceptions that normally result from unrelated code

File: src/compiler/sys.ts (hunk 4 above)
Comment (sheetalkamat) [reply src/compiler/sys.ts:442]: Yes `finds package.json on demand, watches for deletion, and removes them from cache` test verifies that.

File: src/compiler/sys.ts (hunk 3 above)
Comment (sheetalkamat) [reply src/compiler/sys.ts:1086]: Its watcher that handles if file or directory is present on disk and based on that either create file system event based watcher(if exists) or polling watcher (for missing)

Comment (sheetalkamat) [reply src/compiler/sys.ts]: Renamed it to make it clearer

File: src/compiler/sys.ts (hunk 1 above)
Comment (amcasey) [reply src/compiler/sys.ts:452]: Please consider adding a comment to that effect (i.e. it's only expected to be null in the event of a file watcher error).

**Comment by sheetalkamat** [comment]
@amcasey any other feedback or is this good to go.. I would like to get this in sooner than later. Thanks

**Comment by sheetalkamat** [comment]
Just didnt think about it.. i can change that to `UseFsEventsWithFallbackDynamicPolling`

**Comment by DanielRosenwasser** [comment]
I think it depends whether you think one is "riskier" than the other. Fewer people opt themselves into the dynamic polling behavior, right? But we can still get it in in time for the beta if you think it would be the right course of action.

**Comment by sheetalkamat** [comment]
feels like relying on native implementation seems like better option but our dynamic polling is good.. But we dont have experience with its usage in fallbacksetting (eg how does reponse time look with very few files in polling queue etc) so safer bet seems to be `useFsEvents`. i feel like people have not complained about this much given for directories default is `useFsEvents`.

**Comment by DanielRosenwasser** [comment]
That's fair. I do think we should keep an eye out and revisit. My impression is that Windows and macOS, I assume this almost never makes a difference; but on Linux, it will be pretty common to hit the file watching limit, so we may still end up installing a lot of file-watchers and it would be good to improve the situation. Let me know if that's an accurate assessment.

**Comment by sheetalkamat** [comment]
> end up installing a lot of file-watchers and it would be good to improve the situation. Let me know if that's an accurate assessment.

Yes thats true.. also many users just increase limit of native file watchers instead of opting for fallback but then this was before things were not easily configurable. 

Definitely can revisit this at later point.

**Comment by AaronFriel** [comment]
@sheetalkamat have to say I thought the blog post was interesting! I just did a [deep dive on file-watching in Go for Pulumi](https://www.pulumi.com/blog/pulumi-watch-mode-with-rust/) and we ended up using a Rust binary to ship the feature.

Does this implementation of file-watchers use ignore files at all, to reduce the number of system calls needed to monitor a directory containing, e.g., node_modules?

**Comment by DanielRosenwasser** [comment]
Love to find the community of people banging their heads against cross-platform file-watching!

During review of the blog post @sheetalkamat mentioned there's a few specific things in editor scenarios where we watch the `node_modules` directory instead of individual files within `node_modules`. But I might have misunderstood and she can fill in with more context.

**Comment by sheetalkamat** [comment]
As daniel mentioned, we watch node_modules instead of watching each file in editor scenarios. But apart from that when watching node_modules recursively, we have been ignoring files/directories starting with "." in `node_modules`. https://github.com/microsoft/TypeScript/blob/main/src/compiler/sys.ts#L487

**Pull request merged**